### DISCUSSION TITLE
Controlling the CF using UART

### DISCUSSION THREAD

**Discussion by arunbhaarathi** [bitcraze/discussions#129] [Q&A]
Hi,
I am trying to communicate with the crazyflie using a raspberrypi zero with UART. The pi is able to communicate successfully with the CF in UART but only with the CF connected to a computer via a USB cable. My UART connection test code fails if the USB connection is removed. My UART test code is run on the raspberrypi zero and crazyflie is powered with the battery.

Test Code:
```
import logging
import time

import cflib.crtp
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.positioning.motion_commander import MotionCommander

# choose the serial URI that matches the setup serial device
URI = 'serial://ttyAMA0'

# Only output errors from the logging framework
logging.basicConfig(level=logging.ERROR)

if __name__ == '__main__':
    # Initialize the low-level drivers including the serial driver
    cflib.crtp.init_drivers(enable_serial_driver=True)

    with SyncCrazyflie(URI) as scf:
        # We take off when the commander is created
        with MotionCommander(scf) as mc:
            print('Taking off!')
            time.sleep(0.1)
            # We land when the MotionCommander goes out of scope
            print('Landing!')
```

**Comment by evoggy** [comment]
Hi,

Ok, from what I've understood this should work but I haven't tested myself. Can you paste the output from the Raspberry when you try to connect using the UART? Are you sure about the URI? Shouldn't it be ```serial://ttyACM0```?

**Reply by arunbhaarathi** [reply]
When the USB is connected it prints 
"Taking off! 
  Landing!"
When the USB is disconnected, no output or error is printed. The terminal screen is just blank.

**Reply by evoggy** [reply]
Ok, so when you only connect the UART TX/RX pins to the Crazyflie (without the USB connected) it doesn't work. How are you powering the Crazyflie without the USB connected? Battery?

If you want printouts from the python library you need to change the log level to info like this:
```
logging.basicConfig(level=logging.INFO)
```

**Reply by arunbhaarathi** [reply]
Yes, I am powering the Crazyflie with battery. The URI for the raspberrypi I'm using is 'serial://ttyAMA0', I'm sure about this.

I implemented the above mentioned changes. Still it didn't work without the USB.

This is the output I get with the USB connected :
```
INFO:cflib.crazyflie:Callback->Connection initialized[serial://ttyAMA0]
INFO:cflib.crtp.serialdriver:Initialized serial driver.
INFO:cflib.crazyflie:We are connected[serial://ttyAMA0], request connection setup
INFO:cflib.crazyflie:Callback->Connected to [serial://ttyAMA0]
INFO:cflib.crazyflie.platformservice:Protocol  version: 5
WARNING:cflib.crazyflie.toccache:Could not save cache, no writable directory
INFO:cflib.crazyflie:Log TOC finished updating
INFO:cflib.crazyflie.mem:4 memories found
INFO:cflib.crazyflie:Memories finished updating
WARNING:cflib.crazyflie.toccache:Could not save cache, no writable directory
INFO:cflib.crazyflie:Param TOC finished updating
INFO:cflib.crazyflie:Callback->Connection setup finished [serial://ttyAMA0]
Taking off!
Landing!
INFO:cflib.crazyflie:Closing link
INFO:cflib.crazyflie:Callback->Disconnected from [serial://ttyAMA0]

```
With the USB disconnected, this is the output:

```
INFO:cflib.crazyflie:Callback->Connection initialized[serial://ttyAMA0]
INFO:cflib.crtp.serialdriver:Initialized serial driver.
INFO:cflib.crazyflie:We are connected[serial://ttyAMA0], request connection setup

```
The screen just freezes, no output or error. I tried keeping the terminal alive for 20mins, there wasn't a connection timeout either.

**Comment by knmcguire** [comment]
hi @arunbhaarathi, 

You should probably update both the crazyflie-firmware and cflib as we removed the old uart connection with this [PR](https://github.com/bitcraze/crazyflie-firmware/pull/1068) and replaced it by pure CPX.

Mind that [these instructions ](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/)has some mistakes for the kbuild setup so I have updated them in the[ following PR](https://github.com/bitcraze/crazyflie-lib-python/pull/348/files).

With updating both the cflib and the crazyflie-firmware with the kbuild configs, I'm able to somehow recreate your problem, albeit there are some more errors. I've tried the [crazyflie-lib-python example basicSyncLog](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/logging/basiclogSync.py) with changing the following:

```
(41) uri = uri_helper.uri_from_env(default='serial://ttyUSB0')
...
(49)    cflib.crtp.init_drivers(enable_serial_driver=True)
```

And if I don't have an usb but only a ftdi cable to the uart2 connected I see:
```
CPX UART transport
Connecting to UART on /dev/ttyUSB0 @ 576000...
```

and it never connects

and if I do connect an usb cable to the micro usb output + a ftdi cable to the uart2, I see:

```
CPX UART transport
Connecting to UART on /dev/ttyUSB0 @ 576000...
0
0
255
Got start
0
Connected
Creating queue for CPXFunction.CRTP
CRC error!
CRC error!
CRC error!
CRC error!
```

So there seems to be two bugs: 
1- the serial ftdi connection doesn't seem to work independently from an usb cable connection: https://github.com/bitcraze/crazyflie-lib-python/issues/349
2- the CRC error which is created to the check sum check of the message: https://github.com/bitcraze/crazyflie-lib-python/issues/350

Please keep an eye on those to see if they will be fixed. Marcus is on a holiday for a few more weeks so I'm not sure if we can fix this ourselves until then.
